### ISSUE TITLE
Specifying that a window function should only be computed for certain rows

### ISSUE THREAD

**Issue by andrea-s** [ClickHouse/ClickHouse#35056]
I'm not sure whether it is already possible somehow, but it would be nice to be able to indicate that a certain window function should only be calculated for some rows.

Example:

```
SELECT
    count(),
    sum(wf)
FROM
(
    SELECT
        number,
        max(number) OVER (PARTITION BY floor(number / 5)) AS wf
    FROM
    (
        SELECT *
        FROM system.numbers
        LIMIT 100000000.
    )
    ORDER BY number ASC
)
WHERE number = 1

Query id: af97dad0-d585-4eb9-ac49-6a5660679eca

┌─count()─┬─sum(wf)─┐
│       1 │       4 │
└─────────┴─────────┘

1 rows in set. Elapsed: 14.722 sec. Processed 100.03 million rows, 800.21 MB (6.79 million rows/s., 54.36 MB/s.)
```

This query is quite intensive, surely because the window function wf is calculated for all rows before filtering. But it would be convenient to be able to indicate that we only need it calculated where number = 1 in order to skip most of the work.

This is obviously a toy example, but it maps quite nicely to actual clickstream analytics problems (e.g. finding the last x pages visited before a conversion... I don't need the window calculated on all rows since I'm only interested in looking at the conversions, which are significantly less than the pageviews).

**Comment by andrea-s** [comment]
p.s. I _think_ it might be the case that short circuiting will help here once https://github.com/ClickHouse/ClickHouse/issues/19857 is closed, but I'm not entirely sure

**Comment by UnamedRus** [comment]
>  But it would be convenient to be able to indicate that we only need it calculated where number = 1 in order to skip most of the work.

Not exactly.

You want to calculate it on top of all rows which land the same partition (of window function) as number = 1 

So, you need to take partition expression `floor(number / 5)` as WHERE filter:
```
WHERE floor(number / 5) = floor(1 / 5)
```

```
SELECT
    count(),
    sum(wf)
FROM
(
    SELECT
        number,
        max(number) OVER (PARTITION BY floor(number / 5)) AS wf
    FROM
    (
        SELECT *
        FROM system.numbers
        LIMIT 100000000.
    )
    WHERE floor(number / 5) = floor(1 / 5)
    ORDER BY number ASC
)
WHERE number = 1

Query id: 6ea32797-31db-48a4-8cc2-cd1cc8332f9d

┌─count()─┬─sum(wf)─┐
│       1 │       4 │
└─────────┴─────────┘

1 row in set. Elapsed: 0.145 sec. Processed 70.64 million rows, 565.13 MB (487.51 million rows/s., 3.90 GB/s.)
```

**Comment by andrea-s** [comment]
Yes, this is possible in this toy example where I'm _only_ calculating that specific window function, but in general it might not be possible to do so (because I'd have other aggregations or window functions which may have a different filtering scope).

In general having a way to specify that an aggregate window function should only actually compute specific rows would be helpful in a range of real life use cases (we find that we tend to avoid window functions at the moment and prefer self-joins because of the slowness).

Real life example: a dataset with pageview events and transaction events - if I wanted to calculate the latest transaction for each hit I'd be better served (computationally) by generating the aggregate of all transactions and timestamps for each user, and join that to the main event table. But syntactically it's much worse than specifying a window function, we only do that for performance reasons.

Let me know if this sounds like it would be worth investigating and I'll provide an anonymized real life dataset and queries to test the above.

**Comment by UnamedRus** [comment]
> WHERE floor(number / 5) = floor(1 / 5)
Related issue
https://github.com/ClickHouse/ClickHouse/issues/15257

> Yes, this is possible in this toy example where I'm only calculating that specific window function, but in general it might not be possible to do so (because I'd have other aggregations or window functions which may have a different filtering scope).

Ok, so you are asking for FILTER clause for window functions

https://modern-sql.com/feature/filter


```
SELECT
    count(),
    sum(wf)
FROM
(
    SELECT
        number,
        max(number) FILTER(WHERE floor(number / 5) = floor(1 / 5)) OVER (PARTITION BY floor(number / 5)) AS wf
    FROM
    (
        SELECT *
        FROM system.numbers
        LIMIT 100000000.
    )
    ORDER BY number ASC
)
WHERE number = 1
```

So if FILTER clause match specific partitions of WINDOW function or subquery WHERE condition match specific partitions of WINDOW function, ClickHouse can avoid calculation of  all partitions. 

> In general having a way to specify that an aggregate window function should only actually compute specific rows would be helpful in a range of real life use cases (we find that we tend to avoid window functions at the moment and prefer self-joins because of the slowness).

-If combinator or FILTER (WHERE clause)

**Comment by andrea-s** [comment]
Yes - that's correct, syntactically, but the execution is very slow because of https://github.com/ClickHouse/ClickHouse/issues/40781, https://github.com/ClickHouse/ClickHouse/issues/42892

**Comment by marzagheddon** [comment]
I need the same exact feature.
I have a bunch of rows that get logged every n seconds, and i need a boolean that is 1 only if the row in the bunch is different from the one in the previous bunch. Then i need the running total of this boolean.

I implemented it with a CTE, calculating the window function for all the rows in the window but then setting 0 to the ones i didn't need the window function on wrapping the window function in a multiIf
I'm new to CH, so maybe I'll figure it out
